Avoid crashing if traverseUnchangedSubtree can't find the same backingSharingProvider.

mattwoodrow opened the pull request:
#### 20ba7372309cd112039c97c47a566e80de1aefe8
<pre>
Avoid crashing if traverseUnchangedSubtree can&apos;t find the same backingSharingProvider.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276754">https://bugs.webkit.org/show_bug.cgi?id=276754</a>
&lt;<a href="https://rdar.apple.com/131972979">rdar://131972979</a>&gt;

Reviewed by Simon Fraser.

This currently is a nullptr crash if backingProviderCandidateForLayer returns
nullptr.

We should be able to see if the candidate used from the last update is still
available, and use that to avoid the crash (but still render incorrectly).

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::BackingSharingState::previousBackingProviderCandidateForLayer):
(WebCore::RenderLayerCompositor::traverseUnchangedSubtree):

Canonical link: <a href="https://commits.webkit.org/281112@main">https://commits.webkit.org/281112@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd42a15a9f03bdf94c4ed9d6f319f544ec11fa6b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58674 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11160 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62301 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9118 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60803 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45638 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9317 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47477 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6492 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60705 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35565 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50731 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28331 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32308 "Passed tests") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK1-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8122 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54261 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8304 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64004 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2587 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8307 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54799 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2596 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54885 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12999 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2175 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33830 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34916 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36000 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34661 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -453,6 +454,16 @@ auto RenderLayerCompositor::BackingSharingState::backingProviderCandidateForLaye
     return &candidate;
 }
 
+auto RenderLayerCompositor::BackingSharingState::previousBackingProviderCandidateForLayer(const RenderLayer& layer) -> Provider*
```
Should this be `existing` not `previous`? I'm not sure what sense of "previous" is intended; it's not previous in traversal order.

mattwoodrow replied:
I meant 'previous' rendering update, but maybe that's not commonly used terminology here. I can change to `existing` instead.

squelart commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -1416,8 +1427,9 @@ void RenderLayerCompositor::traverseUnchangedSubtree(RenderLayer* ancestorLayer,
 
     if (layer.paintsIntoProvidedBacking()) {
-        auto* provider = backingSharingState.backingProviderCandidateForLayer(layer, *this, overlapMap, layerExtent);
+        auto* provider = backingSharingState.previousBackingProviderCandidateForLayer(layer);
```
Since this could render incorrectly (in situations where we previously crashed), would it be worth considering:
1. Keep backingProviderCandidateForLayer as a first attempt.
2. If nullptr -> debug-only `ASSERT`, then `LOG_ERROR`, and fallback on previousBackingProviderCandidateForLayer.
3. optional: Change to a `RELEASE_ASSERT(provider)` after that, so it will pop out more in crash traces.

mattwoodrow replied:
It's not obvious that doing the two attempts would improve outcomes. If `backingProviderCandidateForLayer` returns a non-null answer, but still different to the previous/existing, then we're going to have incorrect painting.

Logging an error is a good idea, hopefully will be helpful if anyone reports incorrect rendering.

mattwoodrow committed:
Avoid crashing if traverseUnchangedSubtree can't find the same backingSharingProvider.
https://bugs.webkit.org/show_bug.cgi?id=276754
<rdar://131972979>

Reviewed by Simon Fraser.

This currently is a nullptr crash if backingProviderCandidateForLayer returns
nullptr.

We should be able to see if the candidate used from the last update is still
available, and use that to avoid the crash (but still render incorrectly).

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::BackingSharingState::previousBackingProviderCandidateForLayer):
(WebCore::RenderLayerCompositor::traverseUnchangedSubtree):

Canonical link: https://commits.webkit.org/281112@main
